src/server/pps/server/api_server.go

```diff
@@ -593,6 +593,29 @@ func forEachCommitInJob(pachClient *client.APIClient, jobID string, wait bool, c
 	return nil
 }
 
+// For backwards compatibility, we want users who have objects that are not associated with projects
+// to belong to the "default" project. This means that both client and server should essentially treat
+// "default" and "" project names to be equivalent.
+func keepJobBasedOnProjects(job *pps.Job, projectsFilter []string) bool {
```

Comment:
Rather than hardcoding `"default"` use `pfs.DefaultProjectName`.

Also, this function is needlessly complex: you are iterating thrice over the projects filter for each job. Just turn the project list into a `map[string]bool` which returns true for any requested project.

Reply:
I think something like this would work:

```go
type stringFilter map[string]bool

func makeStringFilter(f []string) map[string]bool {
	var ff = make(map[string]bool)
	for _, s := range f {
		ff[s] = true
	}
	return ff
}

func (a *apiServer) ListJob(request *pps.ListJobRequest, resp pps.API_ListJobServer) (retErr error) {
	filter := makeStringFilter(request.GetProjects())
	return a.listJob(resp.Context(), request, func(ji *pps.JobInfo) error {
		if len(filter) > 0 && !filter[ji.GetJob().GetPipeline().GetProject().GetName()] {
			return nil
		}
		return errors.EnsureStack(resp.Send(ji))
	})
}


// ListJobSet implements the protobuf pps.ListJobSet RPC
func (a *apiServer) ListJobSet(request *pps.ListJobSetRequest, serv pps.API_ListJobSetServer) (retErr error) {
	pachClient := a.env.GetPachClient(serv.Context())

	// Track the jobsets we've already processed
	seen := map[string]struct{}{}

	// Return jobsets by the newest job in each set (which can be at a different
	// timestamp due to triggers or deferred processing)
	jobInfo := &pps.JobInfo{}
	filter := makeStringFilter(request.GetProjects())
	err := a.jobs.ReadOnly(serv.Context()).List(jobInfo, col.DefaultOptions(), func(string) error {
		if _, ok := seen[jobInfo.Job.ID]; ok {
			return nil
		}
		seen[jobInfo.Job.ID] = struct{}{}

		jobInfos, err := pachClient.InspectJobSet(jobInfo.Job.ID, request.Details)
		if err != nil {
			return err
		}
		
		var affectsProjects bool
		for _,ji := range jobInfos {
			if len(filter) > 0 && filter[ji.Job.Pipeline.Project.GetName()] {
				affectsProjects = true
			}
		}
		if !affectsProjects {
			return nil
		}

		return errors.EnsureStack(serv.Send(&pps.JobSetInfo{
			JobSet: client.NewJobSet(jobInfo.Job.ID),
			Jobs:   jobInfos,
		}))
	})
	return errors.EnsureStack(err)
}
```

Reply:
> I changed the proto field to a map directly

I think that leaving it as a list is probably cleaner from the API perspective. The fact that it’s a map from project name to boolean seems like an implementation detail, not something meaningful to a client of the API. Are there useful semantics to passing a mapping from a project name to _false_?

Reply:
Changed!